design/src/rfcs/rfc0027_new_home_for_cred_cache_and_builder_api_consistency.md

```diff
@@ -0,0 +1,146 @@
+```rust
+// in sdk_config.rs
+
+impl Builder {
+    // --snip--
+    pub fn credentials_cache(mut self, cache: CredentialsCache) -> Self {
+        self.set_credentials_cache(Some(cache));
+        self
+    }
+    // --snip--
+}
+```
+3. `aws-sdk-s3`: after `aws_sdk_s3::Client::new` has been called with the variable `sdk_config`
+```rust
+// in client.rs
+
+impl Client {
+    // --snip--
+    pub fn new(sdk_config: &aws_types::sdk_config::SdkConfig) -> Self {
+        Self::from_conf(sdk_config.into())
+    }
+    // --snip--
+}
+```
+calls
+```rust
+// in config.rs
+
+impl From<&aws_types::sdk_config::SdkConfig> for Builder {
+    fn from(input: &aws_types::sdk_config::SdkConfig) -> Self {
+        let mut builder = Builder::default();
+        builder = builder.region(input.region().cloned());
+        builder.set_endpoint_resolver(input.endpoint_resolver().clone());
+        builder.set_retry_config(input.retry_config().cloned());
+        builder.set_timeout_config(input.timeout_config().cloned());
+        builder.set_sleep_impl(input.sleep_impl());
+	builder.set_credentials_cache(input.credentials_cache().cloned());
+        builder.set_credentials_provider(input.credentials_provider().cloned());
+        builder.set_app_name(input.app_name().cloned());
+        builder.set_http_connector(input.http_connector().cloned());
+        builder
+    }
+}
+
+impl From<&aws_types::sdk_config::SdkConfig> for Config {
+    fn from(sdk_config: &aws_types::sdk_config::SdkConfig) -> Self {
+        Builder::from(sdk_config).build()
+    }
+}
+```
+What this all means is that `CredentialsCache` needs to be accessible from `aws-config`, `aws-types`, and `aws-sdk-s3` (SDK client crates, to be more generic). We originally assumed that `CredentialsCache` would be defined in `aws-config` along with `LazyCredentialsCache`, but the assumption no longer holds because `aws-types` and `aws-sdk-s3` do not depend upon `aws-config`.
+
+Therefore, we need to find a new place in which to create credentials caches accessible from the aforementioned crates.
+
+Proposed Solution
+-----------------
+We propose to move the following items to a new crate called `aws-credential-types`:
+- All items in `aws_types::credentials` and their dependencies
+- All items in `aws_config::meta::credentials` and their dependencies
+
+For the first bullet point, we move types and traits associated with credentials out of `aws-types`. Crucially, the `ProvideCredentials` trait now lives in `aws-credential-types`.
+
+For the second bullet point, we move the items related to credentials caching. `CredentialsCache` with its `Lazy` variant and builder lives in `aws-credential-types` and `CredentialsCache::create_cahe` will be marked as `pub`. One area where we make an adjustment, though, is that `LazyCredentialsCache` depends on `aws_types::os_shim_internal::TimeSource` so we need to move `TimeSource` into `aws-credentials-types` as well.
```

Comment:
Typo fix:

```suggestion
For the second bullet point, we move the items related to credentials caching. `CredentialsCache` with its `Lazy` variant and builder lives in `aws-credential-types` and `CredentialsCache::create_cache` will be marked as `pub`. One area where we make an adjustment, though, is that `LazyCredentialsCache` depends on `aws_types::os_shim_internal::TimeSource` so we need to move `TimeSource` into `aws-credentials-types` as well.
```

Also had a conversation with @rcoh about `TimeSource` today, and from that, I think we should consider moving it into `aws-smithy-types` so that it can become configurable on `SdkConfig` and `Config`, and allow customers building for WASM to provide time where the standard library can't. This can be future work though.

Reply:
Thanks for letting me know! I am for the idea, and it makes more sense to me to have it in `aws-smithy-types` rather than `aws-credential-types` 👍